File: frontend/src/api/templates.ts
import {
  useMutation,
  useQuery,
  useQueryClient,
} from "@tanstack/react-query"

import { ApiError } from "@/api/auth"

export type Template = {
  id: string
  name: string
  department_id: string
  client_id: string
  discipline_id: string
  created_at: string
  updated_at: string
  deleted_at: string | null
}

export type TemplateListResponse = {
  items: Template[]
  total: number
  limit: number
  offset: number
}

export type TemplateCreate = {
  name: string
  department_id: string
  client_id: string
  discipline_id: string
}

export type TemplateUpdate = {
  name?: string
}

async function jsonOrEmpty(res: Response): Promise<unknown> {
  const text = await res.text()
  if (!text) return {}
  try {
    return JSON.parse(text)
  } catch {
    return {}
  }
}

function detailOf(body: unknown, fallback: string): string {
  if (
    body &&
    typeof body === "object" &&
    "detail" in body &&
    typeof (body as { detail: unknown }).detail === "string"
  ) {
    return (body as { detail: string }).detail
  }
  return fallback
}

async function apiCall<T>(
  url: string,
  init: RequestInit,
  fallback: string,
): Promise<T> {
  const res = await fetch(url, { credentials: "include", ...init })
  if (res.status === 204) return undefined as T
  const body = await jsonOrEmpty(res)
  if (!res.ok) throw new ApiError(detailOf(body, fallback), res.status)
  return body as T
}

const listKey = ["templates"] as const

export function useTemplateList() {
  return useQuery<TemplateListResponse, ApiError>({
    queryKey: listKey,
    queryFn: () =>
      apiCall<TemplateListResponse>(
        "/api/admin/templates?limit=200",
        { method: "GET" },
        "Load failed",
      ),
  })
}

export function useTemplateCreate() {
  const qc = useQueryClient()
  return useMutation<Template, ApiError, TemplateCreate>({
    mutationFn: (body) =>
      apiCall<Template>(
        "/api/admin/templates",
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
        },
        "Create failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: listKey })
    },
  })
}

export function useTemplateUpdate() {
  const qc = useQueryClient()
  return useMutation<
    Template,
    ApiError,
    { id: string; body: TemplateUpdate }
  >({
    mutationFn: ({ id, body }) =>
      apiCall<Template>(
        `/api/admin/templates/${id}`,
        {
          method: "PATCH",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
        },
        "Update failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: listKey })
    },
  })
}

export function useTemplateDelete() {
  const qc = useQueryClient()
  return useMutation<void, ApiError, string>({
    mutationFn: (id) =>
      apiCall<void>(
        `/api/admin/templates/${id}`,
        { method: "DELETE" },
        "Delete failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: listKey })
    },
  })
}

export function useTemplate(id: string | undefined) {
  return useQuery<Template, ApiError>({
    queryKey: ["templates", id],
    queryFn: () =>
      apiCall<Template>(
        `/api/admin/templates/${id}`,
        { method: "GET" },
        "Load failed",
      ),
    enabled: !!id,
  })
}

// ---- field defs ---------------------------------------------------------

export type FieldDef = {
  id: string
  template_id: string
  name: string
  field_type: string
  required: boolean
  // Phase 5.2: surface this field in the project-list peek panel
  // metric grid + the project detail right sidebar's Metrics block.
  is_project_metric: boolean
  order_index: number
  options: { choices: string[] } | null
  created_at: string
  updated_at: string
  deleted_at: string | null
}

export type FieldDefListResponse = { items: FieldDef[]; total: number }

export type FieldDefCreate = {
  name: string
  field_type: string
  required: boolean
  is_project_metric?: boolean
  order_index: number
  options: { choices: string[] } | null
}

const fieldsKey = (tid: string) => ["templates", tid, "fields"] as const

export function useFieldDefs(tid: string | undefined) {
  return useQuery<FieldDefListResponse, ApiError>({
    queryKey: fieldsKey(tid ?? ""),
    queryFn: () =>
      apiCall<FieldDefListResponse>(
        `/api/admin/templates/${tid}/fields`,
        { method: "GET" },
        "Load failed",
      ),
    enabled: !!tid,
  })
}

export function useFieldDefCreate(tid: string) {
  const qc = useQueryClient()
  return useMutation<FieldDef, ApiError, FieldDefCreate>({
    mutationFn: (body) =>
      apiCall<FieldDef>(
        `/api/admin/templates/${tid}/fields`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
        },
        "Create failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: fieldsKey(tid) })
    },
  })
}

export function useFieldDefUpdate(tid: string) {
  const qc = useQueryClient()
  return useMutation<FieldDef, ApiError, { id: string; body: FieldDefCreate }>({
    mutationFn: ({ id, body }) =>
      apiCall<FieldDef>(
        `/api/admin/templates/${tid}/fields/${id}`,
        {
          method: "PATCH",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
        },
        "Update failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: fieldsKey(tid) })
    },
  })
}

export function useFieldDefDelete(tid: string) {
  const qc = useQueryClient()
  return useMutation<void, ApiError, string>({
    mutationFn: (id) =>
      apiCall<void>(
        `/api/admin/templates/${tid}/fields/${id}`,
        { method: "DELETE" },
        "Delete failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: fieldsKey(tid) })
    },
  })
}

// ---- milestone defs -----------------------------------------------------

export type MilestoneDef = {
  id: string
  template_id: string
  name: string
  direction: string
  date_model: string
  order_index: number
  created_at: string
  updated_at: string
  deleted_at: string | null
}

export type MilestoneDefListResponse = { items: MilestoneDef[]; total: number }

export type MilestoneDefCreate = {
  name: string
  direction: string
  date_model: string
  order_index: number
}

const milestonesKey = (tid: string) =>
  ["templates", tid, "milestones"] as const

export function useMilestoneDefs(tid: string | undefined) {
  return useQuery<MilestoneDefListResponse, ApiError>({
    queryKey: milestonesKey(tid ?? ""),
    queryFn: () =>
      apiCall<MilestoneDefListResponse>(
        `/api/admin/templates/${tid}/milestones`,
        { method: "GET" },
        "Load failed",
      ),
    enabled: !!tid,
  })
}

export function useMilestoneDefCreate(tid: string) {
  const qc = useQueryClient()
  return useMutation<MilestoneDef, ApiError, MilestoneDefCreate>({
    mutationFn: (body) =>
      apiCall<MilestoneDef>(
        `/api/admin/templates/${tid}/milestones`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
        },
        "Create failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: milestonesKey(tid) })
    },
  })
}

export function useMilestoneDefUpdate(tid: string) {
  const qc = useQueryClient()
  return useMutation<
    MilestoneDef,
    ApiError,
    { id: string; body: MilestoneDefCreate }
  >({
    mutationFn: ({ id, body }) =>
      apiCall<MilestoneDef>(
        `/api/admin/templates/${tid}/milestones/${id}`,
        {
          method: "PATCH",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(body),
        },
        "Update failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: milestonesKey(tid) })
    },
  })
}

export function useMilestoneDefDelete(tid: string) {
  const qc = useQueryClient()
  return useMutation<void, ApiError, string>({
    mutationFn: (id) =>
      apiCall<void>(
        `/api/admin/templates/${tid}/milestones/${id}`,
        { method: "DELETE" },
        "Delete failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: milestonesKey(tid) })
    },
  })
}

// ---- bulk reorder -------------------------------------------------------

export function useFieldDefReorder(tid: string) {
  const qc = useQueryClient()
  return useMutation<void, ApiError, string[]>({
    mutationFn: (ordered_ids) =>
      apiCall<void>(
        `/api/admin/templates/${tid}/fields/reorder`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ ordered_ids }),
        },
        "Reorder failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: fieldsKey(tid) })
    },
    onError: () => {
      // Roll back optimistic update by refetching.
      qc.invalidateQueries({ queryKey: fieldsKey(tid) })
    },
  })
}

export function useMilestoneDefReorder(tid: string) {
  const qc = useQueryClient()
  return useMutation<void, ApiError, string[]>({
    mutationFn: (ordered_ids) =>
      apiCall<void>(
        `/api/admin/templates/${tid}/milestones/reorder`,
        {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ ordered_ids }),
        },
        "Reorder failed",
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: milestonesKey(tid) })
    },
    onError: () => {
      qc.invalidateQueries({ queryKey: milestonesKey(tid) })
    },
  })
}

File: backend/app/schemas/templates.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from backend.app.db.models import (
    FIELD_TYPES,
    MILESTONE_DATE_MODELS,
    MILESTONE_DIRECTIONS,
    SELECT_FIELD_TYPES,
)

# ---- templates -----------------------------------------------------------


class TemplateCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    department_id: uuid.UUID
    client_id: uuid.UUID
    discipline_id: uuid.UUID


class TemplateUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)


class TemplateOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    department_id: uuid.UUID
    client_id: uuid.UUID
    discipline_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class TemplateListResponse(BaseModel):
    items: list[TemplateOut]
    total: int
    limit: int
    offset: int


# ---- field defs ----------------------------------------------------------


class FieldDefBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    field_type: str
    required: bool = False
    # Phase 5.2: surface this field's value in the at-a-glance contexts
    # (PeekPanel metric grid + project detail right-sidebar Metrics block).
    is_project_metric: bool = False
    order_index: int = 0
    options: dict[str, Any] | None = None

    @model_validator(mode="after")
    def _check_type_and_options(self) -> "FieldDefBase":
        if self.field_type not in FIELD_TYPES:
            raise ValueError(f"unknown field_type: {self.field_type}")
        is_select = self.field_type in SELECT_FIELD_TYPES
        if is_select and self.options is None:
            raise ValueError(f"{self.field_type} requires options")
        if not is_select and self.options is not None:
            raise ValueError(f"{self.field_type} must not have options")
        if is_select:
            choices = self.options.get("choices") if self.options else None
            if not isinstance(choices, list) or not all(
                isinstance(c, str) and c for c in choices
            ):
                raise ValueError(
                    "options must be {'choices': [non-empty strings, ...]}"
                )
            if len(choices) == 0:
                raise ValueError("options.choices must be non-empty")
        return self


class FieldDefCreate(FieldDefBase):
    pass


class FieldDefUpdate(BaseModel):
    """Patch shape — every field optional, but combinations are validated by
    the route handler against the existing row before persisting."""
    name: str | None = Field(default=None, min_length=1, max_length=200)
    field_type: str | None = None
    required: bool | None = None
    is_project_metric: bool | None = None
    order_index: int | None = None
    options: dict[str, Any] | None = None


class FieldDefOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    template_id: uuid.UUID
    name: str
    field_type: str
    required: bool
    is_project_metric: bool
    order_index: int
    options: dict[str, Any] | None
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class FieldDefListResponse(BaseModel):
    items: list[FieldDefOut]
    total: int


# ---- milestone defs ------------------------------------------------------


class MilestoneDefBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    direction: str
    date_model: str
    order_index: int = 0

    @model_validator(mode="after")
    def _check_enums(self) -> "MilestoneDefBase":
        if self.direction not in MILESTONE_DIRECTIONS:
            raise ValueError(f"unknown direction: {self.direction}")
        if self.date_model not in MILESTONE_DATE_MODELS:
            raise ValueError(f"unknown date_model: {self.date_model}")
        return self


class MilestoneDefCreate(MilestoneDefBase):
    pass


class MilestoneDefUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    direction: str | None = None
    date_model: str | None = None
    order_index: int | None = None


class MilestoneDefOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    template_id: uuid.UUID
    name: str
    direction: str
    date_model: str
    order_index: int
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class MilestoneDefListResponse(BaseModel):
    items: list[MilestoneDefOut]
    total: int


# ---- reorder -------------------------------------------------------------


class ReorderRequest(BaseModel):
    ordered_ids: list[uuid.UUID]
